design/each-expression.md

```diff
@@ -0,0 +1,125 @@
+# "Each" Template Expression
+
+## Scenarios
+
+### Scenario: Wrap steps
+
+For example, consider the following template:
+
+```yaml
+parameters:
+  jobs: []
+
+jobs:
+- ${{ each job in parameters.jobs }} # Each job
+  - job: ${{ job.job }}
+    displayName: ${{ job.displayName }}
+    dependsOn: ${{ job.dependsOn }}
+    condition: ${{ job.condition }}
+    strategy: ${{ job.strategy }}
+    continueOnError: ${{ job.continueOnError }}
+    timeoutInMinutes: ${{ job.timeoutInMinutes }}
+    cancelTimeoutInMinutes: ${{ job.cancelTimeoutInMinutes }}
+    container: ${{ job.container }}
+    variables: ${{ job.variables }}
+    pool: ${{ job.pool }}
+    workspace: ${{ job.workspace }}
+    steps:
+      - task: SetupMyBuildTools@1  # Pre steps
+      - ${{ job.steps }}           # Users steps
+      - task: PublishMyTelemetry@1 # Post steps
+        condition: always()
+```
+
+We can shorten the example above, by leveraging an `each` expression to iterate over the job mapping itself:
+
+```yaml
+parameters:
+  jobs: []
+
+jobs:
+- ${{ each job in parameters.jobs }}   # Each job
+  - ${{ each pair in job }}:           # Insert all properties other than "steps"
+      ${{ if ne(pair.key, 'steps') }}:
+        ${{ pair.key }}: ${{ pair.value }}
+    ${{ if job.steps }}:               # Wrap the steps
+      steps:
+      - task: SetupMyBuildTools@1      # Pre steps
+      - ${{ pair.value }}              # Users steps
+      - task: PublishMyTelemetry@1     # Post steps
+          condition: always()
+```
+
+### Scenario: Wrap jobs
+
+For example, consider the following template:
+
+```yaml
+parameters:
+  jobs: []
+
+jobs:
+- job: CredScan                      # Cred scan first
+  pool: MyCredScanPool
+  steps:
+  - task: MyCredScanTask@1
+- ${{ each job in parameters.jobs }} # Then each job
+  - job: ${{ job.job }}
+    displayName: ${{ job.displayName }}
+    ${{ if job.dependsOn }}:         # Inject dependency
+      dependsOn:
+      - ${{ CredScan }}
+      - ${{ job.dependsOn }}
+    ${{ if not(job.dependsOn) }}:
+      dependsOn:
+      - CredScan
+    condition: ${{ job.condition }}
+    strategy: ${{ job.strategy }}
+    continueOnError: ${{ job.continueOnError }}
+    timeoutInMinutes: ${{ job.timeoutInMinutes }}
+    cancelTimeoutInMinutes: ${{ job.cancelTimeoutInMinutes }}
+    container: ${{ job.container }}
+    variables: ${{ job.variables }}
+    pool: ${{ job.pool }}
+    workspace: ${{ job.workspace }}
+    steps: ${{ job.steps }}
+```
```

Comment:
Does this mean this could be written as such (simpler, no need to explicit list out everything)

```
parameters:
  jobs: []

jobs:
- job: CredScan                      # Cred scan first
  pool: MyCredScanPool
  steps:
  - task: MyCredScanTask@1
- ${{ each pair in job }}:           # Insert all properties other than "steps"
    ${{ if ne(pair.key, 'dependsOn') }}:
      ${{ pair.key }}: ${{ pair.value }}
  ${{if job.dependsOn}}
     dependsOn:
      - ${{ CredScan }}
      - ${{ job.dependsOn }}
  ${{ if not(job.dependsOn) }}:
      dependsOn:
      - CredScan
```

Reply:
yep